x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java

```diff
@@ -42,37 +40,60 @@
         StandardOpenOption.CREATE,
         StandardOpenOption.SPARSE };
 
+    /**
+     * Reference counter that counts the number of eviction listeners referencing to this cache file. Once this instance has been evicted
+     * and all listeners notified it makes sure to delete the physical file backing this cache.
+     */
     private final AbstractRefCounted refCounter = new AbstractRefCounted("CacheFile") {
         @Override
         protected void closeInternal() {
-            CacheFile.this.finishEviction();
+            assert assertNoPendingListeners();
+            try {
+                Files.deleteIfExists(file);
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
         }
     };
 
-    private final ReentrantReadWriteLock.WriteLock evictionLock;
-    private final ReentrantReadWriteLock.ReadLock readLock;
-
     private final SparseFileTracker tracker;
     private final String description;
     private final Path file;
 
-    private volatile Set<EvictionListener> listeners;
-    private volatile boolean evicted;
+    private final Set<EvictionListener> listeners = new HashSet<>();
+
+    /**
+     * A reference counted holder for the current channel to the physical file backing this cache file instance.
+     */
+    private static final class FileChannelReference extends AbstractRefCounted {
```

Comment:
By moving access to the file channel to a ref-counted basis and giving the channel its own life-cycle, we can remove all need for locking when dealing with file closing and opening.
Background operations running for index inputs that get closed concurrently are tied to a specific instance of this reference and will simply fail once all references to the channel have been released.

Reply:
Sure done :)